File: 1145-Binary Tree Coloring Game.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def btreeGameWinningMove(self, root, n, x):
        """
        :type root: TreeNode
        :type n: int
        :type x: int
        :rtype: bool
        """
        self.left = 0
        self.right = 0
        def dfs(node):
            if not node:
                return 0
            
            if node.val == x:
                self.left = dfs(node.left)
                self.right = dfs(node.right)
                return 0
            
            return 1 + dfs(node.left) + dfs(node.right)
        
        dfs(root)
        return max(max(self.left, self.right), n-self.left-self.right-1) > n/2

File: JavaScript/1145-Binary Tree Coloring Game.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} n
 * @param {number} x
 * @return {boolean}
 */
var btreeGameWinningMove = function(root, n, x) {
    let leftSide = 0, rightSide = 0;
    
    let dfs = (node) =>{
        if (!node) return 0;
        
        if (node.val === x){
            leftSide = dfs(node.left);
            rightSide = dfs(node.right);
            return 0;
        }
        
        return 1 + dfs(node.left) + dfs(node.right);
    }
    
    dfs(root);
    
    return Math.max( Math.max(leftSide, rightSide), n-leftSide-rightSide-1 ) > n/2;
};